### ISSUE TITLE
installing and executing from non-default target directory

### ISSUE THREAD

**Issue by LTLA** [CGATOxford/UMI-tools#100]
Hi guys,

Thanks for this tool. I'm hoping to use it for some of my analyses, but I'm having some trouble getting it running. I've installed it to a non-default target directory via `pip`, using the `-t XXX` option - this seems to have worked fine. But it's not clear to me what I should be doing to make UMI-tools available on the command line. I presume there's a script I have to add to `bin` somewhere?

Anyway, to get around the above issue, I tried installing from source using `python setup.py install` from the GitHub repository. This also worked fine and added a `umi_tools` script to my `PATH`. However, trying to run the code on real data brings up the following error:

```
Assembler messages:
Fatal error: can't create /home/lun01/.pyxbld/temp.linux-x86_64-2.7/lustre/jmlab/software/lib/python2.7/site-packages/umi_tools/_dedup_umi.o: Read-only file system
Assembler messages:
Fatal error: can't create /home/lun01/.pyxbld/temp.linux-x86_64-2.7/lustre/jmlab/software/lib/python2.7/site-packages/umi_tools/_dedup_umi.o: Read-only file system
Traceback (most recent call last):
  File "/lustre/jmlab/software/bin/umi_tools", line 11, in <module>
    load_entry_point('umi-tools==0.4.2', 'console_scripts', 'umi_tools')()
  File "/lustre/jmlab/software/lib/python2.7/site-packages/umi_tools/umi_tools.py", line 43, in main
    module = imp.load_module(command, file, pathname, description)
  File "/lustre/jmlab/software/lib/python2.7/site-packages/umi_tools/dedup.py", line 286, in <module>
    import umi_methods
ImportError: No module named umi_methods
```

`umi_tools` is trying to build compiled objects on its first run, as mentioned in the installation guide. The error is probably because I'm executing the job on a LSF system, where `/home/lun01/` is not writeable from cluster nodes. In any case, the path of the location for the compiled objects seems mangled; the installation directory `XXX` is `/lustre/jmlab/software/lib/python2.7/site-packages/` (which _is_ actually writeable by the nodes). I don't know much about Cython, but is some adjustment necessary to account for non-default installation directories?

Any guidance would be appreciated.

Thanks,

Aaron

**Comment by IanSudbery** [comment]
Hi Aaron,

Thanks for your question. we thought we had this covered but apparently not! Is there somewhere you can run `umi_tools dedup --help` where you have read/write access to both `/lustre/jmlab/software/lib/python2.7/site-packages/` and `/home/lun01` ? This should only happen on the first run.

**Comment by LTLA** [comment]
Running `umi_tools dedup --help` works fine, and makes an object file at the following location:

```
/home/lun01/.pyxbld/temp.linux-x86_64-2.7/lustre/jmlab/software/lib/python2.7/site-packages/umi_tools/_dedup_umi.o
```

This seems like a rather odd place, though. A better place would be the actual installation directory:

```
/lustre/jmlab/software/lib/python2.7/site-packages/umi_tools/_dedup_umi.o
```

In any case, I'll re-run UMI-tools on my data set now that the objects are built. I'm using the command:

```
umi_tools dedup -I out.bam --paired --output-stats=out.log -S dedup.bam
```

... where `out.bam` is coordinate-sorted.

**Comment by LTLA** [comment]
__Update:__ The good news is that `umi_tools` no longer throws an `ImportError` in the repeated run.

The bad news is that it throws another error:

```
 Traceback (most recent call last):
  File "/lustre/jmlab/software/bin/umi_tools", line 11, in <module>
    load_entry_point('umi-tools==0.4.2', 'console_scripts', 'umi_tools')()
  File "/lustre/jmlab/software/lib/python2.7/site-packages/umi_tools/umi_tools.py", line 46, in main
    module.main(sys.argv)
  File "/lustre/jmlab/software/lib/python2.7/site-packages/umi_tools/dedup.py", line 595, in main
    outfile.write(read)
  File "/lustre/jmlab/software/lib/python2.7/site-packages/umi_tools/umi_methods.py", line 102, in write
    key = read.query_name, read.next_reference_name, read.next_reference_start
  File "pysam/calignedsegment.pyx", line 923, in pysam.calignedsegment.AlignedSegment.next_reference_name.__get__ (pysam/calignedsegment.c:11272)
  File "pysam/calignmentfile.pyx", line 1638, in pysam.calignmentfile.AlignmentFile.getrname (pysam/calignmentfile.c:18168)
  File "pysam/calignmentfile.pyx", line 663, in pysam.calignmentfile.AlignmentFile.get_reference_name (pysam/calignmentfile.c:8913)
ValueError: reference_id -1 out of range 0<=tid<456
```

Presumably this comes about because of unmapped mates.

**Comment by IanSudbery** [comment]
Are you running `group --output-unmapped`?



If so then the unmapped mate of an unmapped mate would be my guess.
Otherwise sounds like you have a funky BAM file.



For the moment I’d run without the `--output-unmapped`. I’ll look into a
more permanent fix.

**Comment by IanSudbery** [comment]
Okay, that was easier than I thought. You could have a go with the code referenced in #101 if you like. Hopefully that should solve that problem. 

As for the weird build location. Believe it or not that is Cython's default build location!

**Comment by LTLA** [comment]
Thanks Ian. I haven't run `umi_tools group` at all; just went straight to `dedup` after alignment/sorting. I'm currently validating the BAM file with _Picard_ to make sure that it's okay; I had already run it through `FixMateInformation` before trying to run `umi_tools`, so I can't imagine there would be much wrong with it. In the meantime, I'll try out the code in #101, hopefully it solves the problem.

Anyway, going back to my original post, I've got two questions:

- If I install via `pip`, what script should I add to my `bin` to be able to use `umi_tools` on my `PATH`?
- How easy would it be for Cython to recognize non-default install locations and to build the object files at those locations instead? I ask because I manage the installed software for our group, and it would be nice to avoid (a) making redundant copies of the object files in everyone's home directory, and (b) having to deal with complaints about `umi_tools` crashing because of the issue I ran into above.

Maybe these have obvious answers, or not - I'm a klutz in Python who flails at the keyboard until it works.

**Comment by IanSudbery** [comment]
Dedup should discard reads with unmapped mates by checking the 0x8 bit of the flag. If that is not set, then the tid of the mate shouldn’t be -1. If your file comes out of Picard with a thumbs up, perhaps you could send me some of it so I can work out what is going on. 

For your other questions:
* the `umi_tools` script is what is known in pythonland as a `console_script`. It is generated on the fly by setup.py (which is executed by pip). If you use `--target` for pip, then the python is installed in that location, but the console_script is built in a location derived from `$PATH`.  Probably the best thing to do in your case is to  `export PYTHONUSERBASE=/lustre/jmlab/software` and then run `pip install` with `--user`. this will install the python module to `/lustre/jmlab/software/lib/pythonX.X/site-packages/umi_tools` and the console_scripts to `/lustre/jmlab/software/bin`. You should then be able to move the scripts to where ever you want them. 
* It almost certainly is possible to do this. In fact I thought we *had* done this. The default location is hard-coded into Cython, so I'm going to have to do some investigating on this one.

**Comment by LTLA** [comment]
Got it (probably). The BAM file contains supplementary alignments (`0x800` set), which are labelled as pairs but do not have `0x8` set or a sensible value of `mtid`. This is because `FixMateInformation` ignores supplementary and secondary alignments during the fixing. I assume that `umi_tools` isn't ignoring these alignments, which causes the observed error.

To give more detail, this is not RNA-seq data. Instead, it's more like a Hi-C experiment, where you get chimeric reads due to sequencing across a ligation junction. The read is thus broken up into several alignments, all of which but the 5'-most one are reported as supplementary. The 5' alignments of each read are most important, so I don't expect `umi_tools` to consider the position of the supplementary alignments during dedupping. 

However, it would be nice if dedupping handled the supplementary alignments gracefully, i.e., discard all supplementaries belonging to discarded read pairs, while keeping those for retained read pair. This is because the supplementary alignments can occasionally be useful in downstream processing, so tossing all of them out prior to dedupping would be a waste.

**Comment by TomSmithCGAT** [comment]
This does sound like the likely source. We currently leave the user to decide whether they want to remove secondary/supplementary reads prior to deduplication. Since each position is considered independently, I'd usually recommend using an input BAM with a single alignment for each read. However, I can see in your case you may wish to retain the supplementary alignments for downstream analyses.

@IanSudbery We could handle this situation by including a `-skip-secondary` option so secondary and supplementary reads are skipped, and a `output-secondary` to output the secondary/supp. reads. The later would require modifying `network.TwoPassPairWriter` to either write out the secondary/supp reads during `write_mates` or else adding a `write_secondary`. Single end reads would also need a new _writer_ class. What do you think?

**Comment by IanSudbery** [comment]
Our philosophy on supplementary and secondary alignments is that what users want to do with them is going to depend entirely on what their application is, how their mapper is using the flags and their existing views on multimapping, non-local alignment etc. Thus we though the best way to deal with these issues was to not deal with them and just treat secondary and supplementary alignments like any other alignment; leaving it up to the user to decide what they wanted to do with them. 

Specifically, your scheme (which I’ll admit would appeal to me as well), would be difficult to implement without storing every supplementary alignment and every the name of every primary alignment that we output in memory until the end of the run. (Because we can’t guarantee anything about ordering, we would need to check when we encountered a supplementary whether we had encounter its primary and if not store it until we did). This is what we do with read pairs, but as 1) we can guarantee that each read1 is matched to exactly one read2 and 2) pairs tend to be close to each other – in generally we only need to store a few hundred to a few thousand reads in memory at once. 

My recommendation would be the following:

Filter out the secondary and supplementary alignment from your file. Run dedup. Use the output of dedup to refill the supplementary alignments. I don’t know what your favorite language is, but a python script to do this would look like:

```python
import pysam
dedup_bam = pysam.AlignmentFile(“deduped.bam”)

keep_reads = set()
for read in dedup_bam.fetch(until_eof=True):
    keep_reads.add(read.query_name)

dedup_bam.close()

supplementary_bam = pysam.AlignmentFile(“original.bam”)
out_bam = pysam.AlignmentFile(“output.bam”, “wb”, template=supplementary_bam)

for read in supplementary_bam.fetch(until_eof=True):
    if not read.is_secondary and read.is_supplementary and read.query_name in keep_reads:
        out_bam.write(read)

out_bam.close()
supplementary_bam.close()
```

You’ll then need to do a `samtools merge` on dedup.bam and output.bam

**Comment by IanSudbery** [comment]
Hi @LTLA, Can you try the code in branch dd9c54a - it should build, compile and install the cython extensions in the UMI-tools directory. 

@TomSmithCGAT Reading around, it seems like pyximport is probably bad practice as it monkey patches core python keywords (import), and can foobar other Cython code running in the same process (not an immediate problem as we have control over everything in the local process at the moment). We could look into distributing the cythonized `_dedup_umi.c` and compile at install time, but I believe this would require users to run `setup.py build_ext`  as well as `setup.py install`, or for us to distribute prebuilt binaries for major OSes. Sounds like a hassle. If and when network.py is replaced with a proper C library, this might become immaterial, because that will have to be compiled and installed in a different way anyway.

**Comment by LTLA** [comment]
The good news is that after removing the supplementary alignments, `umi_tools` has been running happily, so hopefully that issue is solved. The not-so-good news is that the `pyximport_inplace` branch doesn't alter the build location. I did the following:

1. Switched to the `pyximport_inplace` branch in the Git repo.
2. Ran `python setup.py install`.
3. Deleted the `~/.pyxbld` folder.
4. Ran `umi_tools dedup --help`.

This regenerates the object files at the funny location described previously:

```
/home/lun01/.pyxbld/temp.linux-x86_64-2.7/lustre/jmlab/software/lib/python2.7/site-packages/umi_tools/_dedup_umi.o
```

If it helps, I see  `_dedup_umi.pyx` and `_dedup_umi.c` files at:

```
/lustre/jmlab/software/lib/python2.7/site-packages/umi_tools-0.4.2-py2.7.egg/umi_tools/
```

which is where all Python modules go when I use `pip`.

**Comment by IanSudbery** [comment]
Well that is inexplicable and annoying … it works on my system (every
software maintainer’s famous last words eh?).



Can I ask you to do a couple of things for me…



1)      Start a python interpreter with `python`
import the umi_tools network module with `import umi_tools.network`
ask where this is being imported from with `umi_tools.network.__file__`

2)  Open up the `network.py` file in
`/lustre/jmlab/software/lib/python2.7/site-packages/umi_tools-0.4.2-py2.7.egg/umi_tools/network.py`
and let me know how the pyximport.install call looks like (should be
on line 18). It should read `pyximport.install(build_in_temp=False,
inplace=True, build_dir = os.path.dirname(__file__))`

3)      Check for a file called `network.pyc` and delete it if you find it.

**Comment by LTLA** [comment]
_"works on my system"_ - story of my life :)

Anyway, your prompts made me realise that I had pulled down the branch but not the latest commit, which was why `network.py` was actually not up to date. Once I corrected this, the strange stuff at `~/.pyxbld` is no longer generated on the first run of `umi_tools`.

Instead, I see a shared object at the more sensible location of:

```
/lustre/jmlab/software/lib/python2.7/site-packages/umi_tools-0.4.2-py2.7.egg/umi_tools/_dedup_umi.so
```

Success! Cython is still making temporary files at a strange location (note the doubled path):

```
/lustre/jmlab/software/lib/python2.7/site-packages/umi_tools-0.4.2-py2.7.egg/umi_tools/temp.linux-x86_64-2.7/lustre/jmlab/software/lib/python2.7/site-packages/umi_tools-0.4.2-py2.7.egg/umi_tools/_dedup_umi.o
```

... but at least this isn't in my home directory anymore.
